src/execution/as_abi.rs

```diff
@@ -12,33 +12,36 @@ use crate::middlewares::gas_calibration::param_size_update;
 use crate::settings;
 use as_ffi_bindings::{BufferPtr, Read as ASRead, StringPtr, Write as ASWrite};
 use function_name::named;
-use wasmer::Memory;
+use wasmer::{AsStoreMut, AsStoreRef, FunctionEnvMut, Memory};
 
 use super::common::{abi_bail, call_module, create_sc, ABIResult};
 use super::{create_instance, get_module, local_call, MassaModule};
 
 /// Get the coins that have been made available for a specific purpose for the current call.
-pub(crate) fn assembly_script_get_call_coins(env: &ASEnv) -> ABIResult<i64> {
-    sub_remaining_gas(env, settings::metering_get_call_coins())?;
+pub(crate) fn assembly_script_get_call_coins(mut ctx: FunctionEnvMut<ASEnv>) -> ABIResult<i64> {
+    let env = ctx.data().clone();
+    sub_remaining_gas(&env, &mut ctx, settings::metering_get_call_coins())?;
     Ok(env.get_interface().get_call_coins()? as i64)
 }
 
 /// Transfer an amount from the address on the current call stack to a target address.
 #[named]
 pub(crate) fn assembly_script_transfer_coins(
-    env: &ASEnv,
+    mut ctx: FunctionEnvMut<ASEnv>,
     to_address: i32,
     raw_amount: i64,
 ) -> ABIResult<()> {
-    sub_remaining_gas(env, settings::metering_transfer())?;
+
+    let env = ctx.data().clone();
+    sub_remaining_gas(&env, &mut ctx, settings::metering_transfer())?;
     if raw_amount.is_negative() {
         abi_bail!("Negative raw amount.");
     }
     let memory = get_memory!(env);
-    let to_address = &get_string(memory, to_address)?;
+    let to_address = &get_string(memory, &mut ctx, to_address)?;
```

Comment:
the memory is inside the `env` that is inside the `ctx` I think everywhere we could pass only one argument ?

Reply:
function takes `impl AsStoreRef` as input type (in an effort to keep everything as generic as possible I believe) so that won't do

Reply:
Arguments for get_string are of type: Memory & Store (or impl AsStoreRef) and so here we pass the full ctx only because it implements the trait...